[C++ API] Make pImpl easy to use in modules to enable happy reference semantics

goldsborough opened the pull request:
This PR implements hopefully the last iteration of how modules should be defined and constructed in the C++ API, which was discussed a few weeks ago with @ezyang and @ebetica. The most minimal definition of a module is now:

```
struct LinearImpl : public torch::nn::Module {
  struct Options {
    Options(int64_t in, int64_t out);
    TORCH_ARG(int64_t, in);
    TORCH_ARG(int64_t, out);
  } options;

  explicit LinearImpl(Options options);
  Variable forward(Variable input);

  Variable weight_;
  Variable bias_;
};
TORCH_MODULE(Linear);
```

and usage is

```
Linear model(3, 4);
model->forward(...);
auto clone = model->clone();
```

`TORCH_MODULE` defines a class `Linear` which wraps a `std::shared_ptr<LinearImpl>`, as well as providing convenient mechanisms to forward constructor arguments to the `Options` struct. Notice that I wrote `Linear(3, 4)` and not `Linear(LinearOptions(3, 4))`. This makes usage very convenient in the most common case. When default options need be changed, it becomes `Linear(LinearOptions(4, 5).default(1).default(2))`.

The most important file is `torch/csrc/api/include/torch/nn/pimpl.h`. The rest is mostly codemods.

Thoughts, wishes, improvements? @ebetica @ezyang

ezyang commented:
How about an example of keyword only args in the PR description?

goldsborough commented:
@ezyang Do you mean like `Linear(LinearOptions(4, 5).default(1).default(2))`?

ezyang commented on torch/csrc/api/include/torch/nn/pimpl.h:
```diff
@@ -0,0 +1,117 @@
+  /// inside braces, as to construct the constructor argument.
+  template <typename T, typename... Ts>
+  explicit ModuleHolder(T&& t, Ts&&... ts)
+      : impl_(new Contained({std::forward<T>(t), std::forward<Ts>(ts)...})) {}
```
So.... why can't you use `make_shared` here? (I'm sure there is a good reason for splitting the constructors this way, but I am not getting it from the comment.)

goldsborough replied:
`make_shared` takes a template argument `T` and templates cannot infer initializer lists in C++, i.e. `make_shared<T>({...})` does not work.

ezyang commented on torch/csrc/api/include/torch/nn/pimpl.h:
```diff
@@ -0,0 +1,117 @@
+  /// Multi-argument constructor. This constructor is special in that the
+  /// expectation is that the constructor of the contained type takes an object
+  /// that can be constructed with the given arguments. For our modules, this is
+  /// always the `Options` struct. For this reason, the arguments are forwarded
```
I don't think the comments here are clear enough about expected usage. Maybe point to an explicit example.

ezyang reviewed (approved):
But I'd advise getting someone who is not me to once-over the design.

goldsborough commented:
Waiting for @ebetica to sign off

ezyang commented:
@ebetica is in China so it might take some time.

apaszke commented:
Having played with some C++ APIs like QT in the past, I always felt pretty bad when I used some semi-magical macros like `TORCH_MODULE` and `TORCH_ARG`. There are different standard ways of handling things that those macros do, that are more native to the language, and hiding those behind the macros makes it pretty hard to reason about the code. Not mentioning that when something actually goes wrong, then debugging is impossible without knowing how the macro expands. Are we sure this is a good idea? I feel that allowing pretty arbitrary ways of construct and deal with modules is generally a sign of a good API. PyTorch should be a library, not a framework that enforces a particular code structure.

goldsborough commented:
@apaszke thanks for sharing your concerns. I very much agree. I did also structure the `Module` class in such a way that one can write Module's without this pImpl mechanism. For example, it is just as valid to have a module like:
```
struct MyModule : torch::nn::Module {
  MyModule(int in, int out) { ... }
  Variable forward(Variable) { ... }
};
auto module = std::make_shared<MyModule>(4, 5);
Sequential sequential(module); // Works
AnyModule any = module; // Works
```
As long as your module is in a `shared_ptr` and inherits from `torch::nn::Module`, there are pretty much no other constraints. I think this will allow users who don't want to partake in this pImpl scheme to write their own style of modules, which will interact perfectly with the rest of the library.

ebetica reviewed (approved):
LGTM

goldsborough committed:
Created TORCH_MODULE macro

Rewrote Linear

Rewrote Dropout and added default constructor to TORCH_MODULE macro

Turned TORCH_MODULE contens into a proper base class

Added some documentation

Got rid of the old Dropout module

Got rid of the old Embedding module

Got rid of the old BatchNorm module

Got rid of the old Conv module

Fixing optimizers

Rebase

Removed old RNN modules and the TORCH_ATTR macro

Removed temporary P:: namespace

Added cloning behavior to all modules

Got rid of some get() calls

self review nits

Remove noexcept from ModuleHolder methods that can throw

Remove spaces

Add missing override to reset() methods

Added examples to documentation in pimpl.h

goldsborough committed:
Post rebase fixes

goldsborough commented:
@pytorchbot retest this please

goldsborough commented:
@pytorchbot retest this please
